File: sdk-py/examples/ws_ticker.py
"""Subscribe to ticker.BTC and print 60 seconds of price updates.

Run:
    pip install noether-sdk
    python sdk-py/examples/ws_ticker.py http://127.0.0.1:4000
"""

import asyncio
import sys

from noether_sdk.ws import WsClient


async def main() -> None:
    base_url = sys.argv[1] if len(sys.argv) > 1 else "http://127.0.0.1:4000"
    ws_url = base_url.replace("http", "ws").rstrip("/") + "/v1/ws"

    ws = WsClient(ws_url)
    await ws.connect()
    print(f"connected to {ws_url}")

    def on_tick(data: object, channel: str) -> None:
        if isinstance(data, dict):
            print(f"[{channel}] {data.get('asset')} = {data.get('priceFloat')}")

    await ws.subscribe("ticker.BTC", on_tick)
    await asyncio.sleep(60)
    await ws.close()


if __name__ == "__main__":
    asyncio.run(main())

File: sdk-ts/examples/ws-ticker.ts
/**
 * noether-sdk — WebSocket ticker example.
 *
 * Subscribes to ticker.BTC and prints every price update for 60 seconds.
 *
 *   # Node 22+ (native WebSocket)
 *   npx tsx sdk-ts/examples/ws-ticker.ts http://127.0.0.1:4000
 *
 *   # Older Node, pass the `ws` package:
 *   npx tsx sdk-ts/examples/ws-ticker.ts http://127.0.0.1:4000
 */

import { NoetherClient } from '../src/index.js';

async function main() {
  const baseUrl = process.argv[2] ?? 'http://127.0.0.1:4000';
  const client = new NoetherClient({ baseUrl });
  const ws = client.ws({ onLog: (level, msg) => console.log(`[ws:${level}] ${msg}`) });

  await ws.connect();
  console.log('connected; subscribing to ticker.BTC');

  await ws.subscribe('ticker.BTC', (data, channel) => {
    const t = data as { asset: string; priceFloat: number; timestamp: number };
    console.log(`[${channel}] ${t.asset} = ${t.priceFloat} (oracle ts ${t.timestamp})`);
  });

  setTimeout(() => {
    console.log('done; closing');
    ws.close();
    process.exit(0);
  }, 60_000);
}

main().catch((err) => {
  console.error('ws example failed:', err);
  process.exit(1);
});
